Mk3 pf build 106

3d-gussner opened the pull request:
Thanks to @deliopoulos for the idea to split the downloads.

What is new:
- Splitted PF-build in 3 parts:
  - Arduino IDE 1.8.5 will be downloaded from Arduinio.cc
  - Board definition will be downloaded from `Arduino_Boards` Github repository (at this moment  https://github.com/DRracer/Arduino_Boards/tree/master ). A pull request https://github.com/DRracer/Arduino_Boards/pull/1 is pulled. PR https://github.com/DRracer/Arduino_Boards/pull/2 is open
  - Prusa specific library changes are downloaded from Github reopsotory (at this moment my https://github.com/3d-gussner/PF-build-env/releases/tag/1.0.6).
- Added `python` check, as it it used during language build.
- Added Linux 32-bit support (alpha) not tested.
- Added _smart_ detection if Arduino IDE is used with GUI
- Changed exit numbers
  - 1-13 prepare build environment
  - 21-28 prepare compile
  - 31-36 compile

Tested on Windows 10 64-bit using git-bash and Linux Subsystem.

It is possible to use this Arduino IDE in GUI mode.
Please do not update the `Arduino AVR boards` !!! Only supported version 1.6.20. If you do it you have to delete the `PF-build-env-1.0.6` and run the `PF-build.sh` script again.

3d-gussner committed:
PF-build for v1.0.6
What has changed:
- It downloads Arduino IDE 1.8.5 just once. Good for future build enviroment versions
- Uses Prusa Arduino_Boards. At this moment it download my 3d-gussner fork/branch/release. Pull request is made to DRracers his fork ... and finally to Prusa repository
- Minimized PF-build-env to changed files only. At this moment it only \hardware\tools\avr\avr\lib\ldscripts\avr6.xn

Old but updated:
Arduino IDE is portable and this build environment can be used with GUI.
Important preferences are set during istallation.
One thing users need to do is to go to Boardsmanager and search for Prusa and use EINSy board.

3d-gussner committed:
Typo fixes

3d-gussner committed:
Merge remote-tracking branch 'upstream/MK3' into MK3_PF-build_106

3d-gussner committed:
Merge remote-tracking branch 'upstream/MK3' into MK3_PF-build_106

3d-gussner committed:
Added few things to prevent an unstable ...
build environment using Arduino IDE GUI

3d-gussner committed:
Changed BOARD_URL to DRracers respository...
after he pulled my PR https://github.com/DRracer/Arduino_Boards/pull/1

3d-gussner committed:
Typo

3d-gussner committed:
Merge remote-tracking branch 'upstream/MK3' into MK3_PF-build_106

kuhnmarek commented on PF-build.sh:
```diff
@@ -110,74 +157,223 @@ if ! type zip > /dev/null; then
-		wget https://github.com/3d-gussner/PF-build-env/releases/download/Win-$BUILD_ENV/PF-build-env-Win-$BUILD_ENV.zip || exit 4
-		#cp -f ../../PF-build-env/PF-build-env-Win-$BUILD_ENV.zip PF-build-env-Win-$BUILD_ENV.zip || exit4
+		wget https://downloads.arduino.cc/arduino-$ARDUINO_ENV-windows.zip || exit 7
+		echo "$(tput sgr 0)"
```
Line duplicate?

3d-gussner replied:
Don't understand...
`wget` is used 3 times:
- to download Arduino IDE 1.8.5 directly from Arduino web site
- to download Prusa Arduino boards definition
- to download Prusa specific library change in `avr6.xn` to support 2nd language

kuhnmarek replied:
I meant the line
234: echo "$(tput sgr 0)"
235: echo "$(tput sgr 0)"

3d-gussner replied:
Changed

kuhnmarek commented on PF-build.sh:
```diff
@@ -110,74 +157,223 @@ if ! type zip > /dev/null; then
 #Check if build exists and creates it if not
 if [ ! -d "../build-env" ]; then
-    mkdir ../build-env || exit 2
+    mkdir ../build-env || exit 5
```
Could this folder be named so that it's apparent these are downloaded files for building Prusa Firware? For example "PF-build-dl" or something like that...

3d-gussner replied:
Sure everything is possible ... I used `build-env` as the `build.sh` does the same. Wanted to be as close as possible to the `build.sh` script and` PF-build-env` repository by @mkbel

3d-gussner replied:
Changed to `PF-build-dl`

kuhnmarek commented on PF-build.sh:
```diff
@@ -329,7 +525,7 @@ do
 	#Prepare hex files folders
 	if [ ! -d "$SCRIPT_PATH/../Hex-files/FW$FW-Build$BUILD/$MOTHERBOARD" ]; then
-		mkdir -p $SCRIPT_PATH/../Hex-files/FW$FW-Build$BUILD/$MOTHERBOARD || exit 10
+		mkdir -p $SCRIPT_PATH/../Hex-files/FW$FW-Build$BUILD/$MOTHERBOARD || exit 27
```
Similarly to above - can this folder be named for example "PF-build-hex" so it clearly belongs to Prusa Firmware build?

kuhnmarek commented on PF-build.sh:
```diff
@@ -110,74 +157,223 @@ if ! type zip > /dev/null; then
-echo "OS         :" $OS
-echo "OS type    :" $OSTYPE
+echo "Script path :" $SCRIPT_PATH
+echo "OS          :" $OS
```
Unused variable? Yields empty string for all tested scenarios.

3d-gussner replied:
Which variable is unused?
Here my results with `git-bash`
```
$ bash PF-build.sh
Windows 64-bit found

Script path : <secret path>/Prusa-Firmware/MK3_PF-build_106
OS          : Windows_NT
OS type     : msys

Ardunio IDE : 1.8.5
Build env   : 1.0.6
Board       : PrusaResearchRambo
Specific Lib: PrusaLibrary
```

kuhnmarek replied:
Yes, ok. I have overlooked that Windows use $OS. It is empty just on Linux-based systems.

kuhnmarek commented on PF-build.sh:
```diff
@@ -329,7 +525,7 @@ do
 	#Prepare hex files folders
 	if [ ! -d "$SCRIPT_PATH/../Hex-files/FW$FW-Build$BUILD/$MOTHERBOARD" ]; then
-		mkdir -p $SCRIPT_PATH/../Hex-files/FW$FW-Build$BUILD/$MOTHERBOARD || exit 10
+		mkdir -p $SCRIPT_PATH/../Hex-files/FW$FW-Build$BUILD/$MOTHERBOARD || exit 27
```
Can the "Hex-files" folder be renamed to "PF-build-hex" or something similar to indicate it's for Prusa Firmware build?

3d-gussner replied:
Sure it can be done.

kuhnmarek reviewed (commented):
Hi, we have successfully tested this pull request and it works well. Good job! Some minor suggestions I have left in the comments.

3d-gussner commented:
> Hi, we have successfully tested this pull request and it works well. Good job! Some minor suggestions I have left in the comments.

Thanks, and thank you for the feedback and testing it. May I ask on what OS did you test it?
- Windows 10 64-bit
- Windows 10 32-bit
- Linux Subsystem for Windows (Ubuntu)
- Linux (which distribution and 64- or 32-bit)

At this moment I am busy with another upgrade so most things are defined via the Arduino_boards JSON file.
See https://github.com/DRracer/Arduino_Boards/pull/1#issuecomment-514093579
and https://github.com/3d-gussner/Arduino_Boards/tree/Bootloader

Big advantage is that Arduino IDE GUI users should be able to compile the firmware at the end without extra installation steps.

Is there still a reason that Prusa Firmware is locked down to Arduino IDE 1.8.5 and Arduino AVR boards 1.6.20?
Newer versions have bugfixes for MacOS and other things.

I will update the PR to new folder names after get feedback on my answers

DRracer commented:
@3d-gussner we are still on the 1.8.5 just because of continuity for now. We know, that moving from avr-gcc 4.9 to 5.4 makes a considerably shorter code (~1KB total), we are planning to move to new Arduino version in the future. Changing the compiler requires deep testing (new compiler=new bugs and potentially reveals the hidden ones too :) ).

Btw. are you able to implant the avr6.xn link patch into the board definition?

kuhnmarek commented:
> > Hi, we have successfully tested this pull request and it works well. Good job! Some minor suggestions I have left in the comments.
> 
> Thanks, and thank you for the feedback and testing it. May I ask on what OS did you test it?
> 
>     * Windows 10 64-bit
> 
>     * Windows 10 32-bit
> 
>     * Linux Subsystem for Windows (Ubuntu)
> 
>     * Linux (which distribution and 64- or 32-bit)
> 
> 

Sure, no problem. We thank you for the contribution.

Tested on:
Ubuntu 18.04.2 LTS 64-bit ... ok
Ubuntu  14.04 64-bit ... ok
Lubuntu 18.10 32-bit  (here it tried to download 64 bit version of arduino, because it didn't recognize [uname -m]: "i686" ) 
Windows 10 64-bit using git-bash ... ok
Linux Subsystem for Windows 10 (OpenSUSE) - works after adding $OSTYPE = "linux" case in the OS related section

kuhnmarek commented on PF-build.sh:
```diff
@@ -76,32 +88,67 @@
+	fi
+# Linux 32-bit
+elif [ $OSTYPE == "linux-gnu" ]; then
+	if [ $(uname -m) == "i386" ]; then
```
This condition is not executed on 32-bit systems with $OSTYPE == "linux-gnu" - that results to unset Processor variable and failed wget.

3d-gussner replied:
Thanks for the feedback and testing on so many different OS.
Modified the if statement .. hope it detects it now better.

3d-gussner commented:
> @3d-gussner we are still on the 1.8.5 just because of continuity for now. We know, that moving from avr-gcc 4.9 to 5.4 makes a considerably shorter code (~1KB total), we are planning to move to new Arduino version in the future. Changing the compiler requires deep testing (new compiler=new bugs and potentially reveals the hidden ones too :) ).

I get it, as new releases fix old bugs but have the potential to generate new one

> Btw. are you able to implant the avr6.xn link patch into the board definition?

Did you look into my https://github.com/3d-gussner/Arduino_Boards/tree/Bootloader branch?
- Added bootloader
- Modified few things so Arduino AVR boards update should not break it. (not tested if i am correct)
- 

- Arduino IDE downloads the tools from their download website. So i tried to configure out how these are generated and found something.
- These tools are generated by https://github.com/arduino/toolchain-avr in our case https://github.com/arduino/toolchain-avr/tree/4.9.2-atmel3.5.3-arduino2 tag
- The toolchain-avr downloads it from atmel.no and patches are applied.

So at the end we always need to "patch" or download own tools. Found a way to get the tools in the boards definition, but these are unpatched.
I see few options:
1. Generate Prusa specific toolchain-avr with modified/patched `4.9.2-atmel3.5.4-arduino2`. This will generate different download files depending on OS --> Additional overhead
2. Use something like my https://github.com/3d-gussner/PF-build-env/tree/PF-build-env-v2 and patch the environment.
  - Have to look into https://github.com/arduino/Arduino/wiki/Arduino-IDE-1.5-3rd-party-Hardware-specification#pre-and-post-build-hooks-since-ide-165 maybe that would be an option to patch `avr6.xn`

3d-gussner committed:
Merge remote-tracking branch 'upstream/MK3' into MK3_PF-build_106

3d-gussner committed:
Changed Folders and Finding OS version
- Changed Build-env path to "PF-build-dl" as requested in PR https://github.com/prusa3d/Prusa-Firmware/pull/2028
- Changed Hex-files folder to PF-build-hex as requested in PR
- Added Finding OS version routine so supporting new OS should get easier

3d-gussner commented:
@kuhnmarek After testing the changes I updated the PR.

Would be interesting if the Linux 32-bit now recognized.

kuhnmarek committed:
Linux 32-bit recognition fixed

3d-gussner committed:
Merge pull request #59 from kuhnmarek/3d-gussner/MK3_PF-build_106

Linux 32-bit recognition fixed

3d-gussner committed:
Merge remote-tracking branch 'upstream/MK3' into MK3_PF-build_106

kuhnmarek commented:
@3d-gussner 
Thank you for your contribution. We are now updating the documentation for the changes in the build system. Could you also review and, if needed, update the README.md section related to PF-build.sh ?
